woocommerce/woocommerce-ios
# Reprint Shipping Label: reprint action with a coordinator

Part of #2169 

## Why

This PR implements the reprint action to **request printable document for a shipping label**, and then **present the AirPrint dialog** on success. I wanted to make the navigation testable, and created a `ReprintShippingLabelCoordinatorProtocol` that has a set of navigation functions during the reprint flow. Before writing `ReprintShippingLabelCoordinatorProtocol`, I tried using `Coordinator` protocol and callback-based actions for various navigation actions in `ReprintShippingLabelViewController`. But I felt like I didn't need a `UINavigationController` (it's also an implementation detail) and the callbacks with a callback within itself (e.g. `func showPaperSizeSelector`) are quite hard to read (and we can't specify the parameter name). If you have any suggestions on how to make the coordinator/navigation code better, I'm all open to changes!

## Changes

- Moved 5 files without any changes (the ones with 0 diffs)
- Created `ReprintShippingLabelCoordinatorProtocol` with all navigation actions for reprinting a shipping label. `ReprintShippingLabelCoordinator` implements the protocol. Added unit tests for `ReprintShippingLabelCoordinator`
- In `OrderDetailsViewController`, added a lazy var `ReprintShippingLabelCoordinator` that shows reprint UI when tapping on the reprint CTA
- Updated `ReprintShippingLabelViewController` to call `ReprintShippingLabelCoordinatorProtocol` for navigation actions (paper size selector and reprint action so far. there are two more actions in the followup subtasks to present informational modals)
- Added a function to `ReprintShippingLabelViewModel` to request printable data for a shipping label with test cases

## Testing

Prerequisites: the site has [WooCommerce Shipping & Tax plugin](https://docs.woocommerce.com/document/woocommerce-shipping-and-tax/) and there is an order with at least one non-refunded shipping label.

You can set up your store to use a test credit card with instructions in p91TBi-3xD-p2.

- Go to the orders tab
- Tap on an order with at least one non-refunded shipping label --> after syncing, a shipping label package card should be shown
- Tap on "Reprint Shipping Label"
- Choose a different paper size or leave it as the default
- Tap "Print Shipping Label" --> an in-progress modal is shown, then an AirPrint modal is shown with a preview of the shipping label in the selected paper size
- (Optional) If you have a printer that supports AirPrint, feel free to try connecting to the printer in the AirPrint modal 🖨️ 

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/102475568-f97d2d80-4094-11eb-9c37-1d9fd6dde42f.gif)

When there is an error generating the printable document, an error alert is shown (ref p91TBi-3QL-p2#comment-2418):

<img src="https://user-images.githubusercontent.com/1945542/102475856-5d9ff180-4095-11eb-85c5-e96d40a59074.png" width="300" />


Update release notes:
